### ISSUE TITLE
Tensorflow Js predict output zero for all classes

### ISSUE THREAD

**Issue by z2043947** [tensorflow/tfjs#5162]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): built using expo SDK 39 on Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Any Android Phone
- TensorFlow.js installed from (npm or script link): npm 
- TensorFlow.js version: @tensorflow/tfjs": "^2.6.0","@tensorflow/tfjs-react-native": "^0.4.0",
- CUDA/cuDNN version: Keras Model built on CUDA 11.2 in a GPU environment. The model was then converted to layers as well  graph models

**Describe the problem**
Tensorflow Predict out put always zero, irrespective whatever model we choose for any image
**Provide the exact sequence of commands / steps that you executed before running into the problem**
The following function was used to load the model
useEffect(() => {
 async function loadModel(){
  const tfReady = await tf.ready();
  const modelJson = await require("./assets/model/model.json");
  const modelWeight = await require("./assets/model/group1-shard.bin");

   const InsectDetector = await 
      tf.loadLayersModel(bundleResourceIO(modelJson,modelWeight));
    setInsectDetector(InsectDetector)
  }
 }
 loadModel()
}, []); 

The following code was used to load an image for classificaion

const imgBuffer = tf.util.encodeString(Imgb64, 'base64').buffer;
  const raw = new Uint8Array(imgBuffer);
  imageTensor = decodeJpeg(raw);
  const imageTensoresized =  tf.image.resizeBilinear(imageTensor, [224, 224]);
  const offset = tf.scalar(255.0);
  const normalized = (imageTensoresized).div(offset);
  let Result = InsectDetector.predict(normalized.reshape([1,224,224,3]));
  console.log("[+] Insect Class Prediction Result",Result)
  let PrData = Result.dataSync()
  let InsectClass = Result.argMax().dataSync()[0]
  console.log("[+] Insect Class Prediction done Insect class ",PrData,InsectClass)

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.
The out put of predict from console.log
{ kept: false,
      isDisposedInternal: false,
      shape: [ 102 ],
      dtype: 'float32',
      size: 102,
      strides: [],
      dataId: {},
      id: 1162,
      rankType: '1',
      scopeId: 848 },

The ouput of .dataSync() fo 102 classes
 '0': 0,
      '1': 0,
      '2': 0,
      '3': 0,
      '4': 0,
      '5': 0,
      '6': 0,
      '7': 0,
      '8': 0,
      '9': 0,
      '10': 0,
      '11': 0,
      '12': 0,
      '13': 0,
      '14': 0,
      '15': 0,
      '16': 0,
      '17': 0,
      '18': 0,
      '19': 0,
      '20': 0,
      '21': 0,
      '22': 0,
      '23': 0,
      '24': 0,
      '25': 0,
      '26': 0,
      '27': 0,
      '28': 0,
      '29': 0,
      '30': 0,
      '31': 0,
      '32': 0,
      '33': 0,
      '34': 0,
      '35': 0,
      '36': 0,
      '37': 0,
      '38': 0,
      '39': 0,
      '40': 0,
      '41': 0,
      '42': 0,
      '43': 0,
      '44': 0,
      '45': 0,
      '46': 0,
      '47': 0,
      '48': 0,
      '49': 0,
      '50': 0,
      '51': 0,
      '52': 0,
      '53': 0,
      '54': 0,
      '55': 0,
      '56': 0,
      '57': 0,
      '58': 0,
      '59': 0,
      '60': 0,
      '61': 0,
      '62': 0,
      '63': 0,
      '64': 0,
      '65': 0,
      '66': 0,
      '67': 0,
      '68': 0,
      '69': 0,
      '70': 0,
      '71': 0,
      '72': 0,
      '73': 0,
      '74': 0,
      '75': 0,
      '76': 0,
      '77': 0,
      '78': 0,
      '79': 0,
      '80': 0,
      '81': 0,
      '82': 0,
      '83': 0,
      '84': 0,
      '85': 0,
      '86': 0,
      '87': 0,
      '88': 0,
      '89': 0,
      '90': 0,
      '91': 0,
      '92': 0,
      '93': 0,
      '94': 0,
      '95': 0,
      '96': 0,
      '97': 0,
      '98': 0,
      '99': 0,
      '100': 0,
      '101': 0 },

**Comment by pyu10055** [comment]
@z2043947 can you verify if what are the values for the input of the model looks fine:
`normalized .print()`
And ensure the model requires normalization result between [0.0, 1.0]

**Comment by z2043947** [comment]
Thanks Ping

So the value has to be between 0.0 and 1.0? Cannot be 0 or greater than 1?

The attached is the converted tensor from the image.

Tensor
        [[[218, 214, 233],
          [153, 149, 168],
          [154, 153, 171],
          ...,
          [209, 214, 234],
          [204, 207, 228],
          [208, 211, 232]],
    
         [[230, 226, 245],
          [162, 158, 177],
          [157, 156, 174],
          ...,
          [145, 150, 170],
          [142, 145, 166],
          [147, 150, 171]],
    
         [[221, 216, 234],
          [152, 147, 165],
          [149, 145, 164],
          ...,
          [135, 140, 160],
          [131, 134, 155],
          [137, 140, 161]],
    
         ...
         [[243, 243, 243],
          [243, 243, 243],
          [243, 243, 243],
          ...,
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255]],
    
         [[243, 243, 243],
          [243, 243, 243],
          [243, 243, 243],
          ...,
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255]],
    
         [[243, 243, 243],
          [243, 243, 243],
          [243, 243, 243],
          ...,
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255]]]

After dividing by scalar 255 it is all zeros. That does not look right?

Tensor
        [[[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
          ...,
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]],
    
         [[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
          ...,
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]],
    
         [[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
          ...,
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]],
    
         ...
         [[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
          ...,
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]],
    
         [[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
          ...,
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]],
    
         [[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
          ...,
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]]]

regards,
Sankaran

Sankaran Iyer
+61 431 499 867
Sent from Mail for Windows 10

**Comment by vladmandic** [comment]
> After dividing by scalar 255 it is all zeros. That does not look right?

```js
const normalized = imageTensoresized.cast('float32').div(255);
```

**Comment by z2043947** [comment]
> > After dividing by scalar 255 it is all zeros. That does not look right?
> 
> ```js
> const normalized = imageTensoresized.cast('float32').div(255);
> ```

It is not dividing by a scalar that results in zeros. It is the following statement

const imageTensoresized = tf.image.resizeBilinear(imageTensor, [224, 224]); 

Same problem with resizeNearestNeighbor as well.  I am now exploring alternate ways of resizing the image.

**Comment by vladmandic** [comment]
> It is not dividing by a scalar that results in zeros. It is the following statement
> `const imageTensoresized = tf.image.resizeBilinear(imageTensor, [224, 224]);`

You're using `imageTensor = decodeJpeg(raw);` - what is `decodeJpeg`?  
And more importantly, what is the shape of the resulting `imageTensor`? It should be `[1, height, width, 3]`

**Comment by z2043947** [comment]
> > It is not dividing by a scalar that results in zeros. It is the following statement
> > `const imageTensoresized = tf.image.resizeBilinear(imageTensor, [224, 224]);`
> 
> You're using `imageTensor = decodeJpeg(raw);` - what is `decodeJpeg`?
> And more importantly, what is the shape of the resulting `imageTensor`? It should be `[1, height, width, 3]`

import {fetch, bundleResourceIO,decodeJpeg} from "@tensorflow/tfjs-react-native";

decodeJpeg is provided by tensorflow/tfjs-react-native. It returns  tensor3d(buffer, [height, width, channels]) . The tensor seems to be OK. The following is imageTensor,print() added after decodeJpeg

Tensor
        [[[218, 214, 233],
          [153, 149, 168],
          [154, 153, 171],
          ...,
          [209, 214, 234],
          [204, 207, 228],
          [208, 211, 232]],

         [[230, 226, 245],
          [162, 158, 177],
          [157, 156, 174],
          ...,
          [145, 150, 170],
          [142, 145, 166],
          [147, 150, 171]],

         [[221, 216, 234],
          [152, 147, 165],
          [149, 145, 164],
          ...,
          [135, 140, 160],
          [131, 134, 155],
          [137, 140, 161]],

         ...
         [[243, 243, 243],
          [243, 243, 243],
          [243, 243, 243],
          ...,
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255]],

         [[243, 243, 243],
          [243, 243, 243],
          [243, 243, 243],
          ...,
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255]],

         [[243, 243, 243],
          [243, 243, 243],
          [243, 243, 243],
          ...,
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255]]]

Tensor becomes zero after executing this statement 
const imageTensoresized = tf.image.resizeBilinear(imageTensor, [224, 224]);
i
mageTensoresized.print()  added after the statement results in

Tensor
        [[[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
          ...,
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]],

         [[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
          ...,
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]],

         [[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
          ...,
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]],

         ...
         [[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
          ...,
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]],

         [[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
          ...,
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]],

         [[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
          ...,
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]]]

**Comment by z2043947** [comment]
The shape of imageTensoresized is [224,224,3] as required even though it is all zero

**Comment by z2043947** [comment]
I have now ensured that a feed an input tensor between 0 and 1 of shape 1,224,224,3 to the model without using resizeBilinear.  The tensor fed to the mode is as follows:

Tensor
    [[[[0.654902 , 0.8039216, 0.5176471],
       [0.6470588, 0.7960784, 0.509804 ],
       [0.6431373, 0.7921569, 0.5058824],
       ...,
       [0.5568628, 0.7568628, 0.2588235],
       [0.5568628, 0.7568628, 0.2588235],
       [0.4901961, 0.6901961, 0.1921569]],

      [[0.6588235, 0.8078431, 0.5215687],
       [0.6470588, 0.7960784, 0.509804 ],
       [0.6352941, 0.7843137, 0.4980392],
       ...,
       [0.5647059, 0.7647059, 0.2666667],
       [0.5529412, 0.7529412, 0.254902 ],
       [0.4823529, 0.682353 , 0.1843137]],

      [[0.6627451, 0.8117647, 0.5254902],
       [0.6392157, 0.7882353, 0.5019608],
       [0.6196079, 0.7686275, 0.4823529],
       ...,
       [0.5568628, 0.7647059, 0.254902 ],
       [0.5333334, 0.7411765, 0.2313726],
       [0.454902 , 0.6627451, 0.1529412]],

      ...
      [[0.3764706, 0.572549 , 0.145098 ],
       [0.3882353, 0.5843138, 0.1568628],
       [0.3921569, 0.5882353, 0.1529412],
       ...,
       [0.5450981, 0.7372549, 0.3333333],
       [0.5450981, 0.7372549, 0.3333333],
       [0.5450981, 0.7372549, 0.3333333]],

      [[0.3764706, 0.572549 , 0.145098 ],
       [0.3882353, 0.5843138, 0.1568628],
       [0.3921569, 0.5882353, 0.1529412],
       ...,
       [0.5450981, 0.7333333, 0.3411765],
       [0.5450981, 0.7333333, 0.3411765],
       [0.5450981, 0.7333333, 0.3411765]],

      [[0.3803922, 0.5764706, 0.1490196],
       [0.3882353, 0.5843138, 0.1568628],
       [0.3921569, 0.5882353, 0.1529412],
       ...,
       [0.5450981, 0.7333333, 0.3411765],
       [0.5450981, 0.7333333, 0.3411765],
       [0.5450981, 0.7333333, 0.3411765]]]]

But I still get predictions of zero.  I have converted the .h5 file and tried with both layer and graph models. I don know what the issue is now.

Insect Class Prediction Result Tensor {
  "dataId": Object {},
  "dtype": "float32",
  "id": 297,
  "isDisposedInternal": false,
  "kept": false,
  "rankType": "2",
  "scopeId": 1,
  "shape": Array [
    1,
    102,
  ],
  "size": 102,
  "strides": Array [
    102,
  ],
} 0.78978 0.14285714285714285
INFO
13:30
[+] Insect Class Prediction done Insect class  Float32Array [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]

**Comment by z2043947** [comment]
I now verified that I get correct predictions when I turn on remote debugging using the chrome debugger tools. The issue is only in the app's JS on the phone!

**Comment by rthadur** [comment]
Thank for you for confirming , please mention@ to reopen if you still see issue
